### PR TITLE
SCP-2032/SCP-2119 Add governance example contract

### PR DESCRIPTION
**Pull request by sjoerdvisscher**
Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested
- If you updated any cabal files or added Haskell packages:
    - [x] `nix-shell shell.nix --run updateMaterialized` to update the materialized Nix files
- If you changed any Haskell files:
    - [x] `nix-shell shell.nix --run fix-stylish-haskell` to fix any formatting issues

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

**Comment by sjoerdvisscher** [comment]
@j-mueller Things I can't figure out: 

- how to check that somebody owns a token
- how to check the state of the state machine in the tests
- what to do with the voting deadline

File: plutus-use-cases/src/Plutus/Contracts/Governance.hs [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    (GovState oldLaw mph (Just (Voting Proposal{newLaw} votes)), FinishVoting) ->
+        let (Sum ayes, Sum nays) = foldMap (\b -> if b then (Sum 1, Sum 0) else (Sum 0, Sum 1)) votes
+        in if ayes >= requiredVotes -- Enough votes in favor
+            then Just (mempty, State (GovState newLaw mph Nothing) currentValue)
+            else if nays > length initialHolders - requiredVotes -- Enough opposed votes
+                then Just (mempty, State (GovState oldLaw mph Nothing) currentValue)
+                else Nothing -- Not enough votes either way, use cancel-voting to cancel
+
+    (GovState law mph (Just _), CancelVoting) ->
+        Just (mempty, State (GovState law mph Nothing) currentValue)
```
Comment (j-mueller) [inline plutus-use-cases/src/Plutus/Contracts/Governance.hs:166]: if you want to use the deadline here, you can put it into a constraint, replacing `empty` with

```haskell
Constraints.mustValidateIn (Interval.from deadline)
```

File: plutus-use-cases/test/Spec/Governance.hs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import           PlutusTx.Prelude            (ByteString)
+
+tests :: TestTree
+tests =
+    testGroup "governance tests"
+    [ checkPredicate "vote all in favor - SUCCESS"
+        (assertNoFailedTransactions)
+        (doVoting 10 0 1)
+
+    , checkPredicate "vote all againts - SUCCESS"
```
Comment (j-mueller) [inline plutus-use-cases/test/Spec/Governance.hs:40]:
```suggestion
    , checkPredicate "vote all against - SUCCESS"
```

File: plutus-use-cases/src/Plutus/Contracts/Governance.hs [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+{-# LANGUAGE TypeApplications   #-}
+{-# LANGUAGE TypeFamilies       #-}
+{-# LANGUAGE TypeOperators      #-}
+{-# OPTIONS_GHC -fno-ignore-interface-pragmas #-}
+{-# OPTIONS_GHC -fno-omit-interface-pragmas #-}
+{-# OPTIONS_GHC -fno-strictness #-}
+{-# OPTIONS_GHC -fno-specialise #-}
+{-# OPTIONS_GHC -fno-spec-constr #-}
+{-# OPTIONS -fplugin-opt PlutusTx.Plugin:debug-context #-}
+module Plutus.Contracts.Governance (
```
Comment (j-mueller) [inline plutus-use-cases/src/Plutus/Contracts/Governance.hs:18]: missing a comment explaining what the contract does

**Comment by j-mueller** [comment]
> how to check that somebody owns a token

You can use `walletFundsChange` with the token value

> how to check the state of the state machine in the tests

You could write something similar to `Plutus.Contract.Test.valueAtAddress`, except that it gives you UTXOs at the address (using `Wallet.Emulator.Folds.utxoAtAddress`). Then you can check that there's exactly one UTXO with the expected state.

> what to do with the voting deadline

See inline comment

File: plutus-use-cases/src/Plutus/Contracts/Governance.hs [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+data Proposal = Proposal
+    { votingDeadline :: Slot -- TODO: not used yet
+    , newLaw         :: ByteString
+    }
+    deriving stock (Show, Generic)
+    deriving anyclass (ToJSON, FromJSON)
+
+data Voting = Voting
+    { proposal :: Proposal
+    , votes    :: AssocMap.Map PubKeyHash Bool
```
Comment (j-mueller) [inline plutus-use-cases/src/Plutus/Contracts/Governance.hs:59]: I was thinking about this over the weekend and I think there's a problem with storing the public key hashes for votes that have been cast: What prevents you from creating a new address, transfering the token to that address, and then voting a second time?

We probably need to make the voting tokens distinguishable by giving them unique `TokenName`s and then make this a map of `TokenName` to `Bool`.

Comment (sjoerdvisscher) [reply plutus-use-cases/src/Plutus/Contracts/Governance.hs:59]: That makes a lot of sense, thanks!

**Review by j-mueller — APPROVED** [review]
👍

**Pull request merged**